[GOFF] Add writing of section symbols

redstar opened the pull request:
The GOFF format uses symbol definitions to represent sections and symbols. Introducing a section can require up to 3 symbol definitions. However, most of these details are not needed by the AsmPrinter. To mapped from a section (a MCSectionGOFF) to the symbol definitions, a new class called MCGOFFSymbolMapper is used. The same information can also be used by the assembly output, which justifies this centralized approach. Writing the mapped symbols is then straight forward.

redstar committed:
[GOFF] Introduce GOFFWriter class

The GOFFWriter has 2 purposes:
- Simplify resource management
- Enable writing of split DWARF files

It follows the design of the other writer classes. No added functionality at this point.

redstar committed:
[GOFF] Add writing of section symbols

The GOFF format uses symbol definitions to represent sections and
symbols. Introducing a section can require up to 3 symbol definitions.
However, most of these details are not needed by the AsmPrinter.
To mapped from a section (a MCSectionGOFF) to the symbol definitions,
a new class called MCGOFFSymbolMapper is used. The same information
can also be used by the assembly output, which justifies this
centralized approach. Writing the mapped symbols is then straight
forward.

uweigand reviewed (commented):
As a general comment, I'm not really sure why we need the whole `SymbolMapper` indirection.  The various flags for the sections seem basically hard-coded.  Couldn't we just require callers of `getGOFFSection` to pass in the necessary attributes, and then move the whole `SymbolMapper` logic to `initGOFFMCObjectFileInfo` and just have it use the appropriate attributes to begin with?

MaskRay commented on llvm/include/llvm/BinaryFormat/GOFF.h:
```diff
@@ -169,6 +169,91 @@ enum SubsectionKind : uint8_t {
+// in a byte as bit zero. The Flags type helps to set bits in byte according
+// to this numeration order.
+class Flags {
+  uint8_t Val;
```
`= 0` and make the default ctor defaulted `= default;`

redstar replied:
Changed.

MaskRay commented on llvm/include/llvm/BinaryFormat/GOFF.h:
```diff
@@ -169,6 +169,91 @@ enum SubsectionKind : uint8_t {
+};
+
+// Structure for the flag field of a symbol. See
+// https://www.ibm.com/docs/en/zos/3.1.0?topic=formats-external-symbol-definition-record,
```
add a space before the `,`? Otherwise the link doesn't work

redstar replied:
Changed.

MaskRay commented on llvm/include/llvm/BinaryFormat/GOFF.h:
```diff
@@ -169,6 +169,91 @@ enum SubsectionKind : uint8_t {
+
+#undef GOFF_SYMBOL_FLAG
+
+constexpr operator uint8_t() const { return static_cast<uint8_t>(SymFlags); }
```
not indented correctly

redstar replied:
Changed.

MaskRay commented on llvm/test/CodeGen/SystemZ/zos-ppa2.ll:
```diff
@@ -19,7 +19,7 @@
 ; CHECK: L#PPA1_void_test_0:
 ; CHECK:    .long   L#PPA2-L#PPA1_void_test_0       * Offset to PPA2
-; CHECK:    .section    "B_IDRL"
+; CHECK:    .section    ".idrl"
```
Q: Is the previous "B_IDRL" wrong?

redstar replied:
My intention was to have unique, uniform synthetic name for the sections. But with Uli's suggestion this is gone now. In the object file/HLASM file, the name is indeed B_IDRL.

MaskRay commented on llvm/test/MC/GOFF/section.ll:
```diff
@@ -0,0 +1,73 @@
+; RUN: llc <%s --mtriple s390x-ibm-zos --filetype=obj -o - | \
+; RUN:   od -Ax -tx1 -v | FileCheck --ignore-case %s
+; REQUIRES: systemz-registered-target
```
unneeded thanks to
```
% cat llvm/test/MC/SystemZ/lit.local.cfg
if not "SystemZ" in config.root.targets:
    config.unsupported = True
```

redstar replied:
Removed.

Everybody0523 commented on llvm/lib/MC/GOFFObjectWriter.cpp:
```diff
@@ -223,21 +196,222 @@ void GOFFOstream::finalizeRecord() {
+  if (ADA)
+    defineSectionSymbols(*ADA);
+  if (Text)
+    defineSectionSymbols(*Text);
```
Does this need to be inverted? Since you said earlier that "Required order: .text first, then .ada."?

Everybody0523 replied:
Actually, judging by the testcase `section.ll` it seems more likely the above comment is incorrect and this is fine.

If that's the case then nevermind 😅

redstar replied:
I need to update the comment. The LD of the .text section references the PR of the ADA, therefore the ADA needs to come first.

Everybody0523 commented on llvm/lib/MC/GOFFObjectWriter.cpp:
```diff
@@ -223,21 +196,222 @@ void GOFFOstream::finalizeRecord() {
+}
+
+void GOFFWriter::defineSectionSymbols(const MCSectionGOFF &Section) {
+  auto [GOFFSectionData, Found] = SymbolMapper.getSection(Section);
```
Perhaps I'm misunderstanding, but `GOFFSymbolMapper::getSection()` returns <something, true> if the parameter is a "special" section, right? If so perhaps the name "Found" isn't the most accurate description?

redstar replied:
`GOFFSymbolMapper` is gone.

redstar commented on llvm/include/llvm/BinaryFormat/GOFF.h:
```diff
@@ -169,6 +169,91 @@ enum SubsectionKind : uint8_t {
+  GOFF_SYMBOL_FLAG(Mangled, bool, 1, 1)
+  GOFF_SYMBOL_FLAG(Renameable, bool, 2, 1)
+  GOFF_SYMBOL_FLAG(RemovableClass, bool, 3, 1)
+  GOFF_SYMBOL_FLAG(ReservedQwords, ESDReserveQwords, 5, 3)
```
For ED- and PR-type records, this field specifies the amount of space (in multiples of 16-byte quadwords) that the Binder will reserve at the origin of an element.
It's required, and the documentation needs an update.

redstar committed:
Update llvm/include/llvm/MC/MCGOFFSymbolMapper.h

Co-authored-by: Neumann Hon <neumann.hon@ibm.com>

redstar committed:
Update llvm/include/llvm/MC/MCGOFFSymbolMapper.h

Co-authored-by: Neumann Hon <neumann.hon@ibm.com>

redstar committed:
Update llvm/lib/MC/GOFFObjectWriter.cpp

Co-authored-by: Neumann Hon <neumann.hon@ibm.com>

redstar committed:
More updates

redstar commented on llvm/include/llvm/MC/MCGOFFSymbolMapper.h:
```diff
@@ -0,0 +1,148 @@
+// location of data. For example, the equivalent of the ELF .text section is
+// made up of a Section Definition (SD) and a class (Element Definition; ED).
+// The name of the SD symbol depends on the application, while the class has the
+// predefined name C_CODE64.
```
It's AMODE not ILP :-)

redstar committed:
Remove the class GOFFSymbolMapper

redstar committed:
Update formatting

uweigand commented:
> I implemented the suggestion from @uweigand. The GOFF attributes are set directly at the `MCSectionGOFF`, and the `GOFFSymbolMapper` is gone. I still need to update a couple of tests, since now the section names have changed.

Thanks, I actually like this better.   Looking at the current state, I'm wondering if this could be simplified even further: couldn't we simply allocate separate MCSectionGOFF objects for each of the SD / ED / PR (and possibly LD, but that's maybe really rather a label instead of a section?) symbols?

Then each of those could have its own name (there'd probably still have to be a synthesized name to avoid dupliation?) and attributes, and then a "parent" link to identify the containing section.   During emission, we could follow those links to ensure the proper numbering and ordering.   Or even simpler, we might be able to assign the ID numbers already in the MCSectionGOFF, similar to how the UniqueID is handled for MCSectionELF.

Finally, this would allow representing the ADA link explicitly, without having to impose a particular policy in the low-level writer.

Does this make sense or am I missing something here?

uweigand commented on llvm/lib/CodeGen/TargetLoweringObjectFileImpl.cpp:
```diff
@@ -2767,15 +2774,55 @@ MCSection *TargetLoweringObjectFileGOFF::getExplicitSectionGlobal(
   std::string Name = ".gcc_exception_table." + F.getName().str();
-  return getContext().getGOFFSection(Name, SectionKind::getData(), nullptr, 0);
+  constexpr bool Is64Bit = true;
+  constexpr bool UsesXPLINK = true;
```
These are of course always true in the current upstream code base.  I'm not sure it makes sense to add support for features not (currently) usable upstream ...

redstar replied:
My goal was to mark those values somehow, because it is very difficult to consistently find those spots again. But using a constant is enough, so this is gone now.

uweigand commented on llvm/lib/CodeGen/TargetLoweringObjectFileImpl.cpp:
```diff
@@ -2767,15 +2774,55 @@ MCSection *TargetLoweringObjectFileGOFF::getExplicitSectionGlobal(
+  constexpr bool Is64Bit = true;
+  constexpr bool UsesXPLINK = true;
+
+  if (Kind.isBSS() || Kind.isData()) {
```
This also adds `isData` - was that deliberate?

redstar replied:
Yes.

uweigand commented on llvm/lib/MC/GOFFObjectWriter.cpp:
```diff
@@ -239,6 +309,102 @@ class GOFFWriter {
+      GOFFSymbol PR = createGOFFSymbol(Section.getLDorPRName(),
+                                       Section.getPRAttributes(), ED.EsdId);
+      PR.SectionLength = Asm.getSectionAddressSize(Section);
+      if (Section.getName() == ".ada") {
```
I don't think this name will ever match with the latest code.   In any case, this is really not the proper place to put this kind of decision logic ... that should really be handled elsewhere (or more generically), I think.

redstar replied:
Yes, that does not work.

redstar replied:
Changed.

uweigand commented on llvm/lib/CodeGen/TargetLoweringObjectFileImpl.cpp:
```diff
@@ -2759,6 +2760,12 @@ MCSection *TargetLoweringObjectFileXCOFF::getSectionForLSDA(
+void TargetLoweringObjectFileGOFF::getModuleMetadata(Module &M) {
+  // Set the main file name if not set previously by the tool.
+  if (getContext().getMainFileName().empty())
+    getContext().setMainFileName(M.getSourceFileName());
```
Hmm.  This sets the common-code "MainFileName", where it otherwise wouldn't be set.  I'm not sure if this is causing any problems - but it would be a place where we behave differently than others, which is usually better to avoid ...

If it makes sense in general to default MainFileName to SourceFileName when compiling a module, maybe that could and should be done in common code.

Otherwise, could we instead set the name to MainFileName (as default) in initGOFFMCObjectFileInfo, and then override to SourceFileName here (if empty)?   Probably need to hold the Root/Ada names as members of TLOF then.

redstar commented:
> Does this make sense or am I missing something here?

I think that splitting the SD/ED/LD into 3 "section"s implies that a MCSectionGOFF has a fundamentally different semantic than the other MCSectionXXX. This is something I would like to avoid. On the other hand, the SD/ED pair is almost the same as an ELF section, so just putting those 2 into a MCSectionGOFF instance and handling the LD/PR symbol differently makes sense. In HLASM, a section definition looks like
```
HLASMHEL CSECT                        * The SD symbol
HLASMHEL AMODE 64
HLASMHEL RMODE ANY
C_CODE64 CATTR EXECUTABLE,RENT        * The ED symbol; the LD symbol is automatically generated.
         ENTRY MAIN
MAIN     AMODE 64
MAIN     XATTR LINKAGE(XPLINK),SCOPE(LIBRARY),REFERENCE(CODE)
MAIN     DS    0FD                    * A "normal" label, with attributes. Results in another LD symbol
```

When I switch sections, and later want to continue the code section, then I need to repeat both statements in a short form:

```
HLASMHEL CSECT ,
C_CODE64 CATTR ,
```

I thought about this when moving the definitions from the SymbolMapper to the MCObjectFileInfo class, but I got no good idea how to handle the LD or PR symbol. A possible way could be to attach the LD or PR symbol as the begin symbol of a section. I need to try this.

One possible downside is that it makes handling of relocations much more complicated. For the base of a relocation, I need the PR symbol (when I have the SD/ED/PR case) or the LD and ED (in the SD/ED/LD case). Currently, this information is available in the MCSectionGOFF, without the need to chase a list to the correct element.

redstar committed:
More updates

uweigand commented:
> I think that splitting the SD/ED/LD into 3 "section"s implies that a MCSectionGOFF has a fundamentally different semantic than the other MCSectionXXX. This is something I would like to avoid. On the other hand, the SD/ED pair is almost the same as an ELF section, so just putting those 2 into a MCSectionGOFF instance and handling the LD/PR symbol differently makes sense.

Thinking a bit more about this, it looks to me that we should treat SD/ED/PR on the one hand differently from LD (and ER) on the other.  The former identify a *range* of address space and may hold contents of those ranges in the form of text records; the latter identify a single address (and hold no content of their own).

From that perspective, the former correspond to the "section" concept, while the latter correspond to the "symbol" concept.   Now, among the section types SD/ED/PR, GOFF is a bit special in that those are nested - this is somewhat similar to the subsection concept, but it is explicit in the object file format (as opposed to, say, ELF subsections).

It seems to me that modelling that nested section concept explicitly by creating a separate MCSectionGOFF for each of SD, ED, and PR, and linking them as appropriate via a `Parent` pointer (which we actually already have today!), doesn't look too fundamentally different ...   As long as we ensure that text emission happens into the right section (ED or PR as appropriate), this should work fine with common code.

In fact, considering that at some point we want to be able to implement a general HLASM AsmParser, which would require handling any allowed combination of CSECT with multiple CATTR, we should *not* merge SD and ED into a single section.  (Also, by having them separately, we no longer need special treatment of the "root" SD in the writer.)

Finally, having separate MCSession structures for each ESD record may allow using the MCSession::Ordinal field as the ESD ID, which matches its purpose for other object file formats, and which would allow easy resolution of parent (and ADA) section pointers to ESD IDs in the writer.

The LD record, on the other hand, clearly should *not* get a MCSectionGOFF.  Rather, it would make sense for this to be represented as a MCSymbolGOFF.   Specifically, this symbol really represents the implicit section start symbol (which ELF also has!); so it should probably best be emitted not from the section table but from the symbol table.  (MCSection already has a `Begin` symbol - it should be possible to use this for that purpose.)   That would also unify emission of that type of LD record with the other LD records for "normal" symbols.

Attributes associated with the LD record should likewise come from the MCSymbolGOFF.  This would include the ADA section, which means that association no longer needs to be hard-coded in the writer, but can instead set up by codegen as appropriate when defining symbols.  (E.g. this would also allow handling arbitrary user-provided XATTR PSECT attributes in an HLASM AsmParser.)

redstar committed:
Fix failing tests

Changes the `printSwitchToSection()` to be more HLASM-ish, and updates the test cases. That was necessary because the emitted name has changed.
However, zos-landingpad.ll is still failing due to a bug.

redstar committed:
Rework based on reviewer comment.

redstar commented:
I refactored the code in the suggested way. Some notes:

- I section to emit the code/data is either the ED or the PR. The other sections are linked via the `Parent` pointer. `MCGOFFStreamer::changeSection()` makes sure that all sections are registered in the correct order.
- I use the ordinal number +1 of the section as the section number, but still need a counter for the symbols.
- I still need to check that the LD symbols are emitted in the correct order.
- The textual output is currently not correct.

uweigand commented on llvm/lib/MC/GOFFObjectWriter.cpp:
```diff
@@ -239,6 +298,63 @@ class GOFFWriter {
+  if (Section.isED()) {
+    GOFFSymbol ED(Section.getName(), Section.getId(),
+                  Section.getParent()->getId(), Section.getEDAttributes());
+    if (Section.requiresLength())
```
Do we really need a special flag for this?   As I understand it, the length should remain zero for EDs that have PRs inside, i.e. those that use GOFF::ESD_BA_Merge - can't we just check for that?   Also, in those cases no text should ever get emitted into the ED section, so it's size should be zero naturally?

redstar replied:
> so it's size should be zero naturally?

Yes, that's true, and it makes the code easier.

uweigand commented on llvm/lib/MC/GOFFObjectWriter.cpp:
```diff
@@ -239,6 +298,63 @@ class GOFFWriter {
+    GOFFSymbol PR(Section.getName(), Section.getId(),
+                  Section.getParent()->getId(), Section.getPRAttributes());
+    PR.SectionLength = Asm.getSectionAddressSize(Section);
+    if (Section.requiresNonZeroLength()) {
```
This is also unclear to me.   What are the actual rules?  Is is that PR sections of size zero are generally not allowed?  Then why do we need an extra flag here?

uweigand replied:
I see.  I think we should implement this rule then - if a section is the target of an ADA link, it cannot be empty.  That would be preferable to upper layers having to remember setting this flag.  (I guess we could still have a flag internally, but that would get automatically set e.g. by `setADA`.)

Or, in the alternative, should we simply *not* set the ADA link if the target section is empty?  If an ADA section, it cannot actually be used for anything, so do we even need to register it as ADA for a text section?

redstar replied:
> Or, in the alternative, should we simply not set the ADA link if the target section is empty? If an ADA section, it cannot actually be used for anything, so do we even need to register it as ADA for a text section?

That is a simple solution, too. I am not sure if this works with the HLASM output.

redstar replied:
> That is a simple solution, too. I am not sure if this works with the HLASM output.

Ok, I contradict myself. Setting the ADA to null gets around the binder error but other parts assume that there is always an ADA. E.g. when calling an external functions.

uweigand replied:
But if you do call an external function, don't you then have to have actual data in the ADA identifying that external function?   I.e. don't those places that require a non-null ADA also automatically imply there are contents?

redstar replied:
For this specific case I do not know what happens under the hood.
For an externally callable function, a function descriptor is generated in the ADA by the _binder_. From the LLVM point of view, the ADA would still be empty. The simple example is:
```
void empty() {}
```
Here, the ADA is empty. Still, the binder generates the function descriptor. A caller of this function can use a function descriptor in its own compilation unit:
```
	DC	R(empty)
	DC	V(empty)
```
I would expect that I get a binder error on `R(empty)` if no associated data area was given for the code section. But I have not tried.

redstar replied:
I talked to a colleague. Exactly this ^^^ case will not work. The binder gives no warning/error but it will cause problems at runtime because the environment parameter is zero.

uweigand replied:
OK, thanks for following up on this.

uweigand commented on llvm/lib/MC/MCContext.cpp:
```diff
@@ -671,24 +672,48 @@ MCContext::getELFUniqueIDForEntsize(StringRef SectionName, unsigned Flags,
   // Do the lookup. If we don't have a hit, return a new section.
-  auto IterBool =
-      GOFFUniquingMap.insert(std::make_pair(Section.str(), nullptr));
+  auto IterBool = GOFFUniquingMap.insert(std::make_pair(T, nullptr));
```
The way the key is defined means that we there can only be a single instance of a section with the same name and type.  For example, there can only be a single ED with name C_WSA64.   This doesn't look correct to me.

uweigand commented on llvm/lib/MC/MCGOFFStreamer.cpp:
```diff
@@ -26,6 +27,29 @@ GOFFObjectWriter &MCGOFFStreamer::getWriter() {
+  MCContext &Ctx = getContext();
+  // Emit the text section.
+  switchSection(Ctx.getObjectFileInfo()->getTextSection());
+}
```
How is this different from the default implementation of `initSections`?

redstar replied:
Removed, not needed.

uweigand commented on llvm/lib/MC/MCGOFFStreamer.cpp:
```diff
@@ -26,6 +27,29 @@ GOFFObjectWriter &MCGOFFStreamer::getWriter() {
+    return;
+  if (Section->getParent())
+    registerSectionHierarchy(Asm, Section->getParent());
+  registerSectionHierarchy(Asm, Section);
```
Huh?  How is this not just an endless recursion?

redstar replied:
Yes, that's a bug. It should call `Asm.registerSection()` instead.

uweigand commented on llvm/include/llvm/MC/MCContext.h:
```diff
@@ -599,8 +615,23 @@ class MCContext {
+public:
+  MCSectionGOFF *getGOFFSection(SectionKind Kind, StringRef Name,
+                                GOFF::SDAttr SDAttributes,
+                                MCSection *Parent = nullptr);
```
SD can never have a parent, so this should be removed.

uweigand commented on llvm/include/llvm/MC/MCContext.h:
```diff
@@ -599,8 +615,23 @@ class MCContext {
+                                MCSection *Parent = nullptr);
+  MCSectionGOFF *getGOFFSection(SectionKind Kind, StringRef Name,
+                                GOFF::PRAttr PRAttributes,
+                                MCSection *Parent = nullptr);
```
On the other hand, ED and PR must have a parent, so it shouldn't default to nullptr.

uweigand commented on llvm/include/llvm/MC/MCGOFFStreamer.h:
```diff
@@ -16,6 +16,9 @@ namespace llvm {
 
 class MCGOFFStreamer : public MCObjectStreamer {
+  std::string RootSDName;
+  std::string ADAPRName;
```
These are no longer used, I think.

redstar commented on llvm/include/llvm/MC/MCSectionGOFF.h:
```diff
@@ -16,34 +16,95 @@
+  // The attributes of the GOFF symbols.
+  GOFF::SDAttr SDAttributes;
+  GOFF::EDAttr EDAttributes;
+  GOFF::PRAttr PRAttributes;
```
Yes, I was thinking about that (but wanted to get everything to work first).

redstar replied:
Changed.

uweigand commented:
> I use the ordinal number +1 of the section as the section number, but still need a counter for the symbols.

Can't we just set the ordinal to the correct value to begin with?  ELF sets the ordinals while writing out the sections; I think we might as well do the same.  (The point being that if we have already ensured we write sections in the correct order, then we will only ever need to refer to ordinals of sections that have already been written.)  Then we'd be back to a single counter used by the writer for all GOFF symbol records, those that come from sections as well as those that come from symbols.

redstar committed:
More work on reviewer comments

Most comments are resolved except:
- I kept the RootSDSection variable
- The attributes are not yet in a union
- I still need to extend the unique section key for hashmap

redstar commented on llvm/include/llvm/MC/MCSectionGOFF.h:
```diff
@@ -16,34 +16,94 @@
+      break;
+    case GOFF::ESD_ST_PartReference:
+      getParent()->printSwitchToSection(MAI, T, OS, Subsection);
+      OS << Name << " XATTR\n";
```
Yes, true, I need to update my HLASM skills.....

redstar committed:
Emit CSECT/CATTR

This adds the logic to emit the CSECT and CATTR assembler instructions.

redstar committed:
Fix key for hashmap and introduce union

redstar commented on llvm/include/llvm/MC/MCObjectFileInfo.h:
```diff
@@ -230,10 +230,11 @@ class MCObjectFileInfo {
+  MCSection *PPA2ListEDSection = nullptr;
+  MCSection *PPA2ListPRSection = nullptr;
+  MCSection *ADAEDSection = nullptr;
+  MCSection *ADAPRSection = nullptr;
```
Removed.

redstar commented:
> > I use the ordinal number +1 of the section as the section number, but still need a counter for the symbols.
> 
> Can't we just set the ordinal to the correct value to begin with? ELF sets the ordinals while writing out the sections; I think we might as well do the same. (The point being that if we have already ensured we write sections in the correct order, then we will only ever need to refer to ordinals of sections that have already been written.) Then we'd be back to a single counter used by the writer for all GOFF symbol records, those that come from sections as well as those that come from symbols.

At first I was a bit confused about this, because the `MCAssembler` sets the ordinal. Well, it seems that the ELF writer overwrites those values. :-) Sure, I can do the same.

redstar committed:
Use getOrdinal/getIndex as the GOFF symbol id.

redstar committed:
Remove initSections()

Also updates the test to match ideas, to make it easier to move those blocks around.

redstar committed:
Remove RootSDSection

redstar commented on llvm/include/llvm/MC/MCObjectFileInfo.h:
```diff
@@ -230,10 +230,11 @@ class MCObjectFileInfo {
-  MCSection *PPA2Section = nullptr;
-  MCSection *PPA2ListSection = nullptr;
-  MCSection *ADASection = nullptr;
+  MCSection *RootSDSection = nullptr;
```
Removed.

redstar committed:
Fix formatting and assertions

redstar committed:
Update the test case to cover all records in the file.

AidoP commented on llvm/lib/CodeGen/TargetLoweringObjectFileImpl.cpp:
```diff
@@ -2759,6 +2762,29 @@ MCSection *TargetLoweringObjectFileXCOFF::getSectionForLSDA(
+  if (FileName.size() > 1 && FileName.starts_with('<') &&
+      FileName.ends_with('>'))
+    FileName = FileName.substr(1, FileName.size() - 2);
+  DefaultRootSDName = Twine(FileName).concat("#C").str();
```
The default root section should be a private section, a section with a name of 1 space. The source file name affecting ABI is quite surprising and could easily lead to linker errors, especially when taking just the stem. For example, `clang src/a.c src/impl/a.c` would have issues.

Additionally I'm not sure if the file name can be relied on at all here. Is it guaranteed to be provided by a compiler frontend? Does it make sense for the file name to be interpreted as UTF-8 then converted to IBM-1047? It also needs to be normalised to deal with characters that are invalid for symbol names.

redstar replied:
Using a name and setting the binding scope to "section scope" is similar to using " " as name and leaving the binding scope unspecified.
The XLC and Open XL compilers provide a command line option to change this name (XLC: `-qcsect`, `-qnocsect`; Open XL: `-mcsect`, `-mnocsect`). The Open XL compiler defaults to the variant coded here but that can be changed to having a name with binding scope unspecified (`-mcsect=a`) or set to space (`-mnocsect`).
Using `-mcsect=a` results in exactly the problem you describe.
The compiler option will be added later to clang, along with the required code here.

The front end (aka clang) provides this value in the `source_filename` property. All strings in LLVM/clang are in UTF-8 so there is no other choice. The same problem arises for symbols derived from function names etc.

AidoP replied:
Thank you, that's very interesting. The documentation seems to suggest that the binding scope attribute only applies to LDs. Interestingly AMBLIST doesn't seem to display it (N/A is shown).

It seems like there are a few undocumented fields and behaviours being relied upon now. Is there anything being done or are there any plans for IBM to update the doc?

redstar replied:
Yes, my plan is to request an update to the documentation.

uweigand commented on llvm/include/llvm/MC/MCGOFFAttributes.h:
```diff
@@ -0,0 +1,113 @@
+struct EDAttr {
+  bool IsReadOnly = false;
+  GOFF::ESDExecutable Executable = GOFF::ESD_EXE_Unspecified;
+  GOFF::ESDAmode Amode;
```
The whole Amode/Rmode handling is a bit confusing to me.  As I understand the GOFF docs, Amode is supposed to be a *symbol* property (i.e. set on LD and ER records) while Rmode is supposed to be an *element* property (i.e. set on ED records).   So it is unclear what an Amode property on ED or PR is supposed to do, exactly. There also doesn't seem to be a way to specify those with any HLASM command I can see.

redstar replied:
No binding problems, so it seems sage to make this change. That said, amblist shows the Amode on PR and ED symbols.

redstar replied:
Ok, I removed this. However, it feels like that there is an inconsistency in the documentation / implementations. The following HLASM code
```
stdin#C CSECT
C_WSA64 CATTR ALIGN(4),DEFLOAD,NOTEXECUTABLE,PART(a),RMODE(64)
        DC      0X
        END
```
results in `RMODE(64)` set at the ED symbol, and `AMODE(64)` set at the PR symbol.
Setting the Amode on a PR symbol makes sense to me because it is not possible to add a LD symbol to the part - this is causing binder errors. To reference the part from a different compilation unit, I have to use `a`, thus the PR has also some symbol semantics.

uweigand replied:
Just to double-check: your current code does *not* set Amode on the PR symbol.   Is is necessary to do this or not?

redstar replied:
To confirm, setting the Amode is not necessary on PR symbols.

uweigand commented on llvm/include/llvm/MC/MCGOFFAttributes.h:
```diff
@@ -0,0 +1,113 @@
+struct LDAttr {
+  bool IsRenamable = false;
+  GOFF::ESDExecutable Executable = GOFF::ESD_EXE_Unspecified;
+  GOFF::ESDNameSpaceId NameSpace = GOFF::ESD_NS_NormalName;
```
According to the GOFF docs, the NameSpace of LD or PR records must always match the NameSpace of the parent ED record.  If that's true, I don't think it is worthwhile allowing to specify this separately here.

redstar replied:
Yes, that's true. I missed that somehow.

uweigand commented on llvm/include/llvm/MC/MCGOFFAttributes.h:
```diff
@@ -0,0 +1,113 @@
+  bool IsRenamable = false;
+  GOFF::ESDExecutable Executable = GOFF::ESD_EXE_Unspecified;
+  GOFF::ESDNameSpaceId NameSpace = GOFF::ESD_NS_NormalName;
+  GOFF::ESDBindingStrength BindingStrength = GOFF::ESD_BST_Strong;
```
For LD (as opposed to ER), it seems "strong" is the only allowed value here.  Again, if this is true, it doesn't make much sense to specify it.

redstar replied:
This is actually needed for C++ inline functions etc. which can end up in several object files. Also for weak definitions, e.g.:

```
// a.c
#include <stdio.h>

__attribute__((weak)) void fun() {
  printf("Weak fun\n");
}

void feature() {
  fun();
}
```
and
```
#include <stdio.h>

extern void feature();

void fun() {
  printf("Other fun\n");
}

int main(int argc, char *argv[]) {
  feature();
  return 0;
}
```
(example taken from a blog by @MaskRay)

Another case were the documentation needs an update.

redstar replied:
I need to ask for this. I see only WXTRN for weak externals in the HLASM documentation.

uweigand replied:
Like above, right now there doesn't appear to be any code to emit LD symbols to asm output.

uweigand commented on llvm/include/llvm/MC/MCGOFFAttributes.h:
```diff
@@ -0,0 +1,113 @@
+  bool IsRenamable = false;
+  bool IsReadOnly = false; // ???? Not documented.
+  GOFF::ESDExecutable Executable = GOFF::ESD_EXE_Unspecified;
+  GOFF::ESDNameSpaceId NameSpace = GOFF::ESD_NS_NormalName;
```
See above for the NameSpace comment.

redstar replied:
Removed.

uweigand commented on llvm/include/llvm/MC/MCGOFFAttributes.h:
```diff
@@ -0,0 +1,113 @@
+  GOFF::ESDExecutable Executable = GOFF::ESD_EXE_Unspecified;
+  GOFF::ESDNameSpaceId NameSpace = GOFF::ESD_NS_NormalName;
+  GOFF::ESDLinkageType Linkage = GOFF::ESD_LT_XPLink;
+  GOFF::ESDAmode Amode;
```
See above for the Amode comment.

redstar replied:
Removed.

uweigand commented on llvm/include/llvm/MC/MCGOFFAttributes.h:
```diff
@@ -0,0 +1,113 @@
+  GOFF::ESDAmode Amode;
+  GOFF::ESDBindingScope BindingScope = GOFF::ESD_BSC_Unspecified;
+  GOFF::ESDDuplicateSymbolSeverity DuplicateSymbolSeverity =
+      GOFF::ESD_DSS_NoWarning;
```
How would you set this via HLASM?  I wasn't able to find anything in the docs ...

redstar replied:
Me too. I need to ask someone with more HLASM knowledge.

redstar replied:
On the other hand, this is the default value and we do not set it to a different value, thus we do not need to specify it.

uweigand commented on llvm/lib/MC/GOFFObjectWriter.cpp:
```diff
@@ -223,13 +197,95 @@ void GOFFOstream::finalizeRecord() {
+        SymbolType(GOFF::ESD_ST_ElementDefinition) {
+    this->NameSpace = Attr.NameSpace;
+    // TODO Do we need/should set the "mangled" flag?
+    SymbolFlags.setFillBytePresence(1);
```
This sets the fill-byte present flag, but doesn't actually allow to specify any fill byte.  Is this deliberate?

redstar replied:
Only a result of using the default value, but I should be more explicit here.

uweigand replied:
Oh, and one more thing here: by hard-coding this here, we're not emitting any HLASM output.   Should we emit a CATTR FILL(0) then as well?   (But that's only supported on the PART apparently, while the fill byte is supposed to go on the ED?)

redstar replied:
That is a good point, because then we can also emit BSS data using the ORG instruction.
According to the AMBLIST output, the fill byte always ends up at the ED element.

uweigand commented on llvm/lib/MC/MCSectionGOFF.cpp:
```diff
@@ -0,0 +1,145 @@
+               GOFF::ESDExecutable Executable, bool IsReadOnly,
+               StringRef PartName) {
+  if (EmitAmodeAndRmode && Amode != GOFF::ESD_AMODE_None) {
+    OS << ParentName << " AMODE ";
```
As above, reading the HLASM docs, the "AMODE" command in GOFF mode should refer to a *symbol* name, not a section name.   This may be a bit confusing due the section start symbol with the same name as the section - I think we should actually emit the AMODE for *that* instead.

(There is some HLASM backward compat mode for AMODE on a section name, but that also actually uses some implicitly generated section start symbol instead.)

redstar replied:
>  I think we should actually emit the AMODE for that instead.
I (or Tony) will do when emitting the labels.

uweigand commented on llvm/lib/MC/MCSectionGOFF.cpp:
```diff
@@ -0,0 +1,145 @@
+  }
+  if (EmitAmodeAndRmode && Rmode != GOFF::ESD_RMODE_None) {
+    OS << ParentName << ' ';
+    emitRMode(OS, Rmode, /*UseParenthesis=*/false);
```
The plain RMODE in GOFF mode HLASM also seems to be only for backward-compat purposes and sets the mode on an implicitly generated B_TEXT class.  I don't think we ever use that class at all - do we really want to set this?  We do need the "CATTR RMODE" which sets the mode on our actual class.

uweigand commented on llvm/lib/MC/MCSectionGOFF.cpp:
```diff
@@ -0,0 +1,145 @@
+    if (!Emitted) {
+      emitCATTR(OS, ED->getName(), ED->getParent()->getName(), !SDEmitted,
+                PRAttributes.Amode, getParent()->EDAttributes.Rmode,
+                PRAttributes.Alignment, getParent()->EDAttributes.LoadBehavior,
```
LoadBehavior and Rmode are only supported on ED, not PR.   Should we even emit those with the CATTR PART at all, or only with the base CATTR for the ED?

redstar replied:
The HLASM statement
```
C_WSA64 CATTR ALIGN(4),DEFLOAD,NOTEXECUTABLE,RMODE(64),PART(a)
```
generates both the ED symbol `C_WSA64` and the PR symbol `a`.
At the ED symbol, Rmode and LoadBehaviour are set.
At the PR symbol, Amode (derived from RMode) and the LoadBehaviour are set. The latter looks like a HLASM bug. I am not aware of a way to defined the class (ED) and the part (PR) separately.

uweigand replied:
The HLASM docs contain the somewhat cryptic statement: "Binding attributes assigned to the class are also assigned to the part." where is not really defined anywhere (I can see) what exactly "binding attributes" mean.

Do we need to replicate the HLASM behavior of setting Amode and LoadBehavior or not?

uweigand commented on llvm/include/llvm/MC/MCGOFFAttributes.h:
```diff
@@ -0,0 +1,113 @@
+// Attributes for ED symbols.
+struct EDAttr {
+  bool IsReadOnly = false;
+  GOFF::ESDExecutable Executable = GOFF::ESD_EXE_Unspecified;
```
According to the GOFF docs, while the executable flag is supported on ED, it's only purpose is to act as default for the executable flags in all LDs defined in this ED.   Assuming we set the flag in the LDs, do we also need it in the ED?

redstar replied:
There may be no LD. For example, the `B_IDRL` section has only SD and ED.

redstar replied:
Well, while my statement is true it also means there is no way to refer to such data.
I checked, it's not required.

uweigand commented on llvm/lib/MC/MCSectionGOFF.cpp:
```diff
@@ -0,0 +1,145 @@
+      emitCATTR(OS, ED->getName(), ED->getParent()->getName(), !SDEmitted,
+                PRAttributes.Amode, getParent()->EDAttributes.Rmode,
+                PRAttributes.Alignment, getParent()->EDAttributes.LoadBehavior,
+                PRAttributes.Executable, PRAttributes.IsReadOnly, Name);
```
Do we also need to set CATTR PRIORITY ?

There are some other PR properties that seem to require XATTR to be set via HLASM, in particular XATTR SCOPE, LINKAGE, and potentially REFERENCE(CODE/DATA).   (I'm not sure how the latter interacts with CATTR [NON]EXECUTABLE, however ... Also, the GOFF docs seem to imply that there is no executable property on PRs, but that may just be wrong ...)

redstar replied:
Yes, PRIORITY is needed.
I also emit the XATTR now. I emit REFERENCE(CODE/DATA), but I test if it is needed. It does not interact with CATTR [NON]EXECUTABLE but may interact with the EXECUTABLE/DATA property set at the external symbol referring to this part. In the past I got binder errors when when those were not matching but I have to check if this is true for PR symbols.

uweigand commented on llvm/lib/MC/MCSectionGOFF.cpp:
```diff
@@ -0,0 +1,145 @@
+      ED->Emitted = true;
+      Emitted = true;
+    } else
+      OS << ED->getName() << " CATTR ,\n";
```
Why the comma at the end?

redstar replied:
My understaning is that it is sometimes done to get past the operand parsing in HLASM.
However, I just realized that the part name needs to be emitted, otherwise the CATTR statement produces an error.

redstar committed:
Reuse NameSpace from ED in PR/LD

redstar committed:
Also set FillByteValue.

redstar committed:
- Amode only for LD
- Add sorting key
- Change generation of Rmode/Amode

redstar committed:
Emit XATTR

redstar committed:
Remove comma and add part name.

redstar committed:
Fix test case due to removed comma

redstar committed:
Remove DuplicateSymbolSeverity since it is not required.

MaskRay commented on llvm/include/llvm/MC/MCContext.h:
```diff
@@ -599,8 +600,18 @@ class MCContext {
 
-  MCSectionGOFF *getGOFFSection(StringRef Section, SectionKind Kind,
-                                MCSection *Parent, uint32_t Subsection = 0);
+private:
```
instead of adding a `private`, just move the `getGOFFSection` to the existing private part.

redstar replied:
Moved the code.

MaskRay commented on llvm/lib/MC/MCGOFFStreamer.cpp:
```diff
@@ -26,6 +27,23 @@ GOFFObjectWriter &MCGOFFStreamer::getWriter() {
 
+namespace {
+// Make sure that all section are registered in the correct order.
+void registerSectionHierarchy(MCAssembler &Asm, MCSectionGOFF *Section) {
```
LLVM style prefers `static` to anonymous namespace for internal functions.

redstar replied:
Remove the anonymous namespace, and added `static`.

MaskRay commented on llvm/lib/MC/MCGOFFStreamer.cpp:
```diff
@@ -26,6 +27,23 @@ GOFFObjectWriter &MCGOFFStreamer::getWriter() {
 
+namespace {
+// Make sure that all section are registered in the correct order.
+void registerSectionHierarchy(MCAssembler &Asm, MCSectionGOFF *Section) {
```
LLVM style prefers `static` to anonymous namespace for internal functions.

redstar replied:
Remove the anonymous namespace, and added `static`.

MaskRay commented on llvm/lib/MC/MCSectionGOFF.cpp:
```diff
@@ -0,0 +1,141 @@
+using namespace llvm;
+
+namespace {
+void emitCATTR(raw_ostream &OS, StringRef Name, GOFF::ESDRmode Rmode,
```
use `static` https://llvm.org/docs/CodingStandards.html#restrict-visibility

redstar replied:
Remove the anonymous namespace, and added `static`.

MaskRay commented on llvm/test/MC/GOFF/section.ll:
```diff
@@ -0,0 +1,112 @@
+; RUN: llc <%s --mtriple s390x-ibm-zos --filetype=obj -o - | \
```
`llc < %s` does not need `-o -`. .ll tests in llvm/test/MC is probably not a good convention we'd recommend. Move to llvm/test/CodeGen/SystemZ, perhaps under a directory zos/ ?

redstar replied:
Moved the tests.

redstar committed:
Change code as recommended by MaskRay

redstar committed:
Move .ll test cases

redstar committed:
Add another test case, and remove -o - in the run line
Also fixes the alignment for data.

MaskRay reviewed (commented):
LGTM. As I'm not familiar with z/OS, my review focused on its compatibility with the current MC infrastructure. (My internet access will be limited between April 20th and May 4th, which may cause delays in my response time.)

Thanks for reimplementing the getPPA1/PPA2 stuff that puzzled me :)

redstar committed:
Remove Executable from the ED element

The value would serve as default if not specified on LD or PR. However, we can make sure that it is always specified on LD and PR, s not need to have it on ED.

redstar committed:
Remove unused header file

uweigand commented on llvm/include/llvm/MC/MCGOFFAttributes.h:
```diff
@@ -0,0 +1,106 @@
+// Attributes for PR symbols.
+struct PRAttr {
+  bool IsRenamable = false;
+  bool IsReadOnly = false; // ???? Not documented.
```
Does it ever make sense for this value to differ from the ED value?  Or is this one of those attributes that should be copied from the element to the part?   It doesn't seem possible to specific differing values in HLASM ...

redstar replied:
I checked again, and it seems we are only using the default value. I remove it.

uweigand commented on llvm/include/llvm/MC/MCGOFFAttributes.h:
```diff
@@ -0,0 +1,106 @@
+  GOFF::ESDExecutable Executable = GOFF::ESD_EXE_Unspecified;
+  GOFF::ESDLinkageType Linkage = GOFF::ESD_LT_XPLink;
+  GOFF::ESDBindingScope BindingScope = GOFF::ESD_BSC_Unspecified;
+  GOFF::ESDAlignment Alignment = GOFF::ESD_ALIGN_Byte;
```
Same question for the alignment, can it ever differ from the ED alignment?   How would you specify this in HLASM?

redstar replied:
The requirement for the alignment is that the aligment of ED must be equal or higher than the alignment of PR. However, to my knowledge it is not possible to share the ED element among several PR elements (because the SD must have the same name as the PR). Which means that both alignments are the same.
In HLASM, the alignment is given with the CATTR statement. If the CATTR defines no PART, then the alignment is set at the ED element, otherwise the same value is set at the ED and PR element.
I remove the alignment in the PR.

redstar committed:
Remove ReadOnly attribute for PR elements

redstar committed:
Make alignment of PR always the same as ED

redstar committed:
Fix test cases

There were too many zeroes in the address.

redstar committed:
Lift fill byte value to higher layer.

redstar committed:
Fix formatting.

MaskRay commented:
3





之

redstar commented:
@MaskRay Sorry, your comment is basically empty. I guess a GitHub problem?

MaskRay commented:
> @MaskRay Sorry, your comment is basically empty. I guess a GitHub problem?

Sorry... Could be my accidentally pushing a comment to a wrong PR..

uweigand reviewed (commented):
Looks generally good to me now, still a few comments inline.

redstar committed:
Update comment.

redstar committed:
Remove constants.

redstar committed:
Rebase and fix compile errors  due to refactoring.

uweigand reviewed (approved):
This version looks good to me now, thanks!
